STORM-1187 Support windowing based on tuple ts

arunmahadevan opened the pull request:
Support for doing window calculations based to tuple timestamps and handle
out of order events based on time lag.

rfarivar commented on storm-core/src/jvm/backtype/storm/windowing/TimeEvictionPolicy.java:
```diff
@@ -21,21 +21,30 @@
 public class TimeEvictionPolicy<T> implements EvictionPolicy<T> {
-    private final long duration;
+    private final int windowLength;
+    protected Long currentTs;
```
What does currentTs represent? May I suggest renaming to something more representative?

arunmahadevan replied:
Its the reference time for the window calculation, based on which events are expired from the window. Will rename to `referenceTime`.

d2r commented:
From the travis-ci run, I see two errors from WaterMarkEventsGeneratorTest:

```
Running backtype.storm.windowing.WaterMarkEventGeneratorTest

Tests run: 4, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.651 sec <<< FAILURE! - in backtype.storm.windowing.WaterMarkEventGeneratorTest

testTrackTwoStreams(backtype.storm.windowing.WaterMarkEventGeneratorTest)  Time elapsed: 0.13 sec  <<< ERROR!

java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:635)
	at java.util.ArrayList.get(ArrayList.java:411)
	at backtype.storm.windowing.WaterMarkEventGeneratorTest.testTrackTwoStreams(WaterMarkEventGeneratorTest.java:90)

testTrackSingleStream(backtype.storm.windowing.WaterMarkEventGeneratorTest)  Time elapsed: 0.191 sec  <<< ERROR!

java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:635)
	at java.util.ArrayList.get(ArrayList.java:411)
	at backtype.storm.windowing.WaterMarkEventGeneratorTest.testTrackSingleStream(WaterMarkEventGeneratorTest.java:62)
```

arunmahadevan commented:
@d2r It seems to be due to timing issues since the unit tests relied on Thread.sleep(). I've removed sleep and updated `WaterMarkEventGeneratorTest` to trigger manually.

Will also review other unit tests for similar issues.

d2r commented:
> @d2r It seems to be due to timing issues since the unit tests relied on Thread.sleep(). I've removed sleep and updated WaterMarkEventGeneratorTest to trigger manually.
> 
> Will also review other unit tests for similar issues.

OK, sounds good to me.  Thanks.

HeartSaVioR commented:
@arunmahadevan I've merge #925, so please upmerge. Maybe you can ignore master's changeset and take #900.

arunmahadevan commented:
@HeartSaVioR I've upmerged.

Parth-Brahmbhatt commented on docs/documentation/Windowing.md:
```diff
@@ -126,6 +126,96 @@ Time duration based tumbling window that tumbles after the specified time durati
+can also be specified which indicates the max time limit for tuples with out of order timestamps. 
+
+E.g. If the lag is 5 secs and a tuple `t1` arrived with timestamp `06:00:05` no tuples may arrive with tuple timestamp earlier than `06:00:00`. If a tuple
+arrives with timestamp 05:59:59 after `t1` and the window has moved past `t1`, it will be treated as a late tuple and not processed. 
```
Lets also document how users can find out number of discarded tuples? In many cases it may also be useful to provide a handler for tuples being discarded but I am fine with not including that in this patch.

Parth-Brahmbhatt commented on storm-core/src/jvm/backtype/storm/windowing/WaterMarkEventGenerator.java:
```diff
@@ -0,0 +1,110 @@
+     */
+    private long computeWaterMarkTs() {
+        long ts = Long.MIN_VALUE;
+        // only if some data has arrived on each input stream
```
why do we have to wait for data to arrive on each input stream?

arunmahadevan replied:
The minimum of the latest event timestamps across all input streams (minus the lag) is considered as the watermark timestamp. This is so that if events from one of the streams is delayed more, we don't treat all events from that stream as late events.

Parth-Brahmbhatt commented:
Overall I am +1.

satishd commented on storm-core/test/jvm/backtype/storm/windowing/WaterMarkEventGeneratorTest.java:
```diff
@@ -0,0 +1,100 @@
+import static org.junit.Assert.*;
+
+/**
+ * Unit tests for {@link WaterMarkEventGeneratorTest}
```
Minor typo: Unit tests for {@link WaterMarkEventGenerator} instead of Unit tests for {@link WaterMarkEventGeneratorTest}

satishd commented on storm-core/src/jvm/backtype/storm/windowing/EvictionPolicy.java:
```diff
@@ -39,4 +63,12 @@
+     *
+     * @param context
+     */
+    void setContext(Object context);
```
May want to introduce class/interface EvictionPolicyContext which may have opaque structure for now. This can be done later as this PR is waiting for long.

satishd commented:
Overall LGTM, +1

harshach commented on storm-core/src/jvm/backtype/storm/Config.java:
```diff
@@ -1697,19 +1697,45 @@
+     * field is not present in the incoming tuple, a java.lang.IllegalArgumentException will be thrown.
+     */
+    @isString
+    public static final String TOPOLOGY_BOLTS_TUPLE_TIMESTAMP_FIELD_NAME = "topology.bolts.tuple.timestamp.field.name";
```
any reason for this config to have BOLT in them , can't we name it as topology.tuple.timestamp.field.name

arunmahadevan replied:
@harshach BOLT was added to emphasize that this is a bolt specific configuration (used for windowed bolts) similar to the other params above this.

harshach commented:
+1

harshach commented:
Thanks @arunmahadevan merged into master.
